### PR TITLE
DAOS-9013 dfuse: skip open counter check in ioil_fini()

### PR DESCRIPTION
**Pull request by wangshilong**
If process quit without calling close(), it will make open
counter missed to decrease, daos_pool_disconnect() will return
EBUSY since ioil_shrink_cont() was skipped.

Signed-off-by: Wang Shilong <shilong.wang@intel.com>

### CODE REVIEW

**Comment by ashleypittman** [comment]
I would expect dfs_umount() to fail if there are open handles, is this not the case?

**Comment by wangshilong** [comment]
> > I would expect dfs_umount() to fail if there are open handles, is this not the case?
> 
> No, when testing, i saw dfs_umount() quit without problems.

I am not sure, i think problem is even process did not call 'close()' explictly, but when process exit, resources will be freed any way, and dfs_umount() should be able to exit properly.

But for IL case, it decrease open counter only when close() called, so that might be reason that such tricky happen?

**Comment by johannlombardi** [comment]
Looks like something important to fix. Could we add a regression test?

File: src/client/dfuse/il/int_posix.c [hunk 1]
```diff
@@ -110,12 +110,12 @@ ioil_shrink_pool(struct ioil_pool *pool)
 }
 
 static int
-ioil_shrink_cont(struct ioil_cont *cont, bool shrink_pool)
+ioil_shrink_cont(struct ioil_cont *cont, bool shrink_pool, bool skip_count_check)
```
Comment (ashleypittman) [inline src/client/dfuse/il/int_posix.c:113]: I think I would prefer this was called force.
```suggestion
ioil_shrink_cont(struct ioil_cont *cont, bool shrink_pool, bool force)
```

**Review by ashleypittman — COMMENTED** [review]
I'm honestly surprised this works, as I put before that if files remained open then dfs_umount() would fail, presumably this isn't ref counted so we're actually leaking dfs handles in this case?  The difference between dfs handles and pool handles though is that dfs handles are local and only in memory (and a leak on shutdown isn't that bad) where as pool handles consume server resources so are best if closed properly?

On the other hand we're using glocal2local to share the dfuse pool handle from with the interception library so at the minute even the pool handle that the interception library is using is a local-only entity isn't it?

**Comment by wangshilong** [comment]
> Looks like something important to fix. Could we add a regression test?

I tried to add one, but it looks not easy to add one, i tried something like this:

LD_PRELOAD=./install/lib64/libioil.so python -c "open('/mnt/client/1.txt', 'r')"

It did not work, only way working now is try to write using c codes(calling open without close()) and then

call something like: LD_PRELOAD=./install/lib64/libioil.so ./a.out

I did not get idea where to integrate this to current test codes.

**Comment by wangshilong** [comment]
> I'm honestly surprised this works, as I put before that if files remained open then dfs_umount() would fail, presumably this isn't ref counted so we're actually leaking dfs handles in this case? The difference between dfs handles and pool handles though is that dfs handles are local and only in memory (and a leak on shutdown isn't that bad) where as pool handles consume server resources so are best if closed properly?
> 
> On the other hand we're using glocal2local to share the dfuse pool handle from with the interception library so at the minute even the pool handle that the interception library is using is a local-only entity isn't it?

It looks to me that we could not avoid all leaks in such case, but trying best we can and avoid errors messages looks better..

**Comment by ashleypittman** [comment]
> > Looks like something important to fix. Could we add a regression test?
> 
> I tried to add one, but it looks not easy to add one, i tried something like this:
> 
> LD_PRELOAD=./install/lib64/libioil.so python -c "open('/mnt/client/1.txt', 'r')"
> 
> It did not work, only way working now is try to write using c codes(calling open without close()) and then
> 
> call something like: LD_PRELOAD=./install/lib64/libioil.so ./a.out
> 
> I did not get idea where to integrate this to current test codes.

We don't have a way to do that now, I'm working on it in https://github.com/daos-stack/daos/pull/7286 however.

**Comment by johannlombardi** [comment]
Please inspect.

**Review by ashleypittman — APPROVED** [review]

**Review by mchaarawi — APPROVED** [review]

**Comment by johannlombardi** [comment]
Please back port to release/2.0.

**Pull request merged**